File: Leetcode/dynamic_programming/1043_PartitionArrayForMaximumSum/Solution.py
class Solution:
    def maxSumAfterPartitioning(self, A: List[int], K: int) -> int:
        n=len(A)
        dp=[0]*(n+1)
        for i in range(1, n+1):
            curMax=float('-inf')
            for j in range(1, min(i,K)+1):
                curMax=max(curMax, A[i-j])
                dp[i]=max(dp[i], dp[i-j]+curMax*j)
        return dp[n]

File: Leetcode/dynamic_programming/1043_PartitionArrayForMaximumSum/Solution.cpp
class Solution {
public:
    int maxSumAfterPartitioning(vector<int>& A, int K) {
        int n = A.size();
        vector<int> dp(n+1, 0);
        for(int i = 1; i<n+1; i++){
            int cur_max = INT_MIN;
            for(int k = 1; k<=min(i, K); k++){
                cur_max = max(cur_max, A[i-k]);
                dp[i] = max(dp[i], dp[i-k] + cur_max * k);
            }
        }
        return dp[n];
    }
};
